### PR TITLE
Generator cleanup

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Generalize and improve generation of unary functions.

- Fix bug which prevented from comparison functions for `DataArray` to work, even though they were generated.
- Allow generating things only from `dataset` upwards, for functions which have non-trivial definitions in `variable`.
- Allow generating functions that do not support `out` args.
- Remove dead code.
- This adds support for `less`, `greater`, ... , `sin`, `cos`, ... , `values`, and `variances` for `DataArray`.

To do in future PRs:
- Remove irrelevant tests.
- Test templates instead.
- Generate wrappers on Python side.
- Use for even more functions.

### CODE REVIEW

File: python/scipp.cpp [hunk 1]
```diff
@@ -28,7 +28,11 @@ void init_unary(py::module &);
 void init_units_neutron(py::module &);
 void init_variable(py::module &);
 
+void init_generated_comparison(py::module &);
 void init_generated_math(py::module &);
+void init_generated_trigonometry(py::module &);
+void init_generated_util(py::module &);
+void init_generated_special_values(py::module &);
```
Comment (jl-wynen) [inline python/scipp.cpp:35]: Should there be an umbrella `init_generated` function that the generator builds from these individual functions?

File: CMake/scipp-util.cmake [hunk 2]
```diff
@@ -21,10 +21,22 @@ function(scipp_function template category function_name)
...
+  else()
+    set(GENERATE_OUT "true")
+  endif()
+
+  if(NOT SCIPP_FUNCTION_SKIP_VARIABLE)
+    configure_in_module("variable" ${function_name})
+  endif()
   configure_in_module("dataset" ${function_name})
   set(src ${NAME}.cpp)
   configure_file(templates/python_${template}.cpp.in python/${src})
```
Comment (jl-wynen) [inline CMake/scipp-util.cmake:42]: This is still generating bindings for variables even if `SKIP_VARIABLE` is set. Could there be situations where a `Dataset` function and binding are trivial but both `Variable` function _and_ binding need special attention?

**Review by jl-wynen — COMMENTED** [review]
Looks good.

Have you compared compile times? I'm a bit afraid that adding a lot of small files blows up the time because the headers have to be re-parsed a lot. Maybe putting all generated functions (at least of one category) into one file would help in case this is really an issue.

File: python/scipp.cpp (hunk 1 above)
Comment (SimonHeybrock) [reply python/scipp.cpp:35]: Yes I agree, I think I will try to add something like that soon.

File: CMake/scipp-util.cmake (hunk 2 above)
Comment (SimonHeybrock) [reply CMake/scipp-util.cmake:42]: Generating Python bindings for variables is definitely deliberate here.

The only special attention I can think of right now is that we generally support `out` args for variable functions, but not for data array functions. This is handled in the template using an `if constexpr`.

Of course we still have (and probably will always have) numerous cases that are not covered by the templates (neither for variable, for data arrays, nor for bindings), so we do them by hand.

**Comment by SimonHeybrock** [comment]
> Looks good.
> 
> Have you compared compile times? I'm a bit afraid that adding a lot of small files blows up the time because the headers have to be re-parsed a lot. Maybe putting all generated functions (at least of one category) into one file would help in case this is really an issue.

I have not, but I think it is (or was when first introduced) actually solving (parts) of the opposite problem: We had certain files with many uses of `transform` that too very long to compile (we had previously manunally split up such files) and caused bad load balance when compiling with multiple threads. If re-parsing headers is an issue we can consider using pre-compiled headers.

**Pull request merged**